### PR TITLE
Adding new layout controllers, tweaking Index files, adding Routers f…

### PR DESCRIPTION
**Pull request by matthewjohnston4**
…or new URLs, and adding a typedef. Part 2/3 of newsite.

Depends on https://github.com/hhvm/user-documentation/pull/19

### CODE REVIEW

File: src/APIIndex.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      $type,
+    );
+    return $index[$type];
+  }
+
+  public static function getFileForAPI(
+    string $type,
+    string $api,
+    ?string $method,
+  ) {
```
Comment (JoelMarcey) [inline src/APIIndex.php]: return type? probably a `: string`

File: src/APIIndex.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      $type,
+    );
+    invariant(
+      $index[$type]->containsKey($api),
+      'Type %s does not contain reference %s',
+      $type,
+      $api,
+    );
+    invariant(
+      isset($index[$type][$api]['path']),
```
Comment (JoelMarcey) [inline src/APIIndex.php]: I think we can use `array_key_exists` on shapes, right?

Maybe `array_key_exists('path', $index[$type][$api])`?

File: src/APIIndex.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      'API %s of type %s does not contain a path',
+      $api,
+      $type,
+    );
+    if ($method !== null) {
+      invariant(
+        isset($index[$type][$api]['methods']),
```
Comment (JoelMarcey) [inline src/APIIndex.php]: same.

File: src/APIIndex.php (hunk 2 above)
Comment (SiebelsTim) [inline src/APIIndex.php]: > I think we can use array_key_exists on shapes, right?

shapes are just a parser alias to array if I remember correctly. 

I think `isset()` is considered evil on a typechecking standpoint, so better not use it..

Comment (JoelMarcey) [inline src/APIIndex.php]: Definitely evil in strict. Yeah, I can get around to changing.

Comment (matthewjohnston4) [inline src/APIIndex.php]: I tried using ->containsKey and got typing errors, but array_key_exists might work

**Pull request merged**